### DISCUSSION TITLE
License changes

### DISCUSSION THREAD

**Discussion by jtauber** [bible-technology/scripture-burrito#276] [Archive]
Currently the `copyright` property consists of the following sub-properties:

- `licenses`
- `publicDomain`
- `shortStatementPlain`
- `shortStatementRich`
- `fullStatementPlain`
- `fullStatementRich`

The final four take IETF-language-indexed strings, the `Plain` ones being plaintext and the `Rich` ones being XHTML.

`publicDomain` is just a boolean.

`licenses` is an array of objects each of with either has a `url` linking to the license on the web or a `ingredient` giving the path within the burrito to the license file.

## Discussion

1. Presumably the reason `licenses` is an array is to handle different licenses for different ingredients within the burrito but there is no mechanism for indicating which license applies to what (although potentially the fullStatements could express that information.
2. Do we really want to inline the fullStatements or should they be pointers to ingredients as well? This has the advantages of not cluttering the JSON and allowing for arbitrary mimetypes (XHTML, Markdown, etc)
3. Why do we sometimes link to ingredients and sometimes rely on an ingredient role? Why shouldn't we use an ingredient role for `license` rather than point to it here?
4. Do we really need a Plain/Rich distinction?

**Comment by jonathanrobie** [comment]
> Do we really want to inline the fullStatements or should they be pointers to ingredients as well? This has the advantages of not cluttering the JSON and allowing for arbitrary mimetypes (XHTML, Markdown, etc)

I think it would be helpful to have `licenses` contain a list of `license`, which each can contain either (1) a reference to a license statement that is an ingredient in the burrito, or (2) a URL that points to a license.

**Reply by joelthe1** [reply]
Also, I can see how people might want multiple files/URLs as part of their licenses (e.g. dual-licensed, long/short, .md/.html, etc).

**Comment by jtauber** [comment]
Straw proposal:

* we keep `licenses` as is
* we keep `publicDomain` as is
* we replace `shortStatementPlain` and `shortStatementRich` with an array `shortStatements` each of whose items contains a mimetype and a language (so you could have a plain text, markdown, HTML, etc in one or more languages)
* we introduce an ingredient role for `fullStatement`

**Reply by jag3773** [reply]
Also add an optional `lang` field on ingredients. If not used, the assumption is that the ingredient using the language specified at the project level.

**Reply by jag3773** [reply]
Set a character limit on shortStatements to 500.

**Comment by jtauber** [comment]
In summary:
 - drop `fullStatementPlain`
 - drop `fullStatementRich`
 - replace `shortStatementPlain` and `shortStatementRich` with `shortStatements` which is an array of objects with `statement`, `lang`, and `mimetype` properties; `statement` is required and has a maxLength of 500.
 - add `lang` property to ingredients
 - add `copyrightStatement` ingredient role to replace `fullStatementX` properties

now see PR #279
